Item_Give and OnReceiveItem updates

Malkierian opened the pull request:
`OnReceiveItem` now takes a `GetItemEntry` parameter to allow access to modIndex (futureproofing for mods).

All access to that hook now feeds in the appropriate parameter.

`Randomizer_Item_Give` now triggers `OnReceiveItem`.

Shop item giving now also implements mod ID tracking to properly feed into `OnReceiveItem` after pending sales go through.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/587778745.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/587778746.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/587778747.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/587778749.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/587778750.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/587778751.zip)
<!--- section:artifacts:end -->

Malkierian committed:
Changed OnReceiveItem hook to pass GetItemEntry, which required the following changes:

Reworked the references to it in `z_parameter` to call a single function with the item return and hook call in it for ease of editing.

Modified the pendingSale functionality to set and pass a modIndex value as well, as mod items (like randomizer) in shops still use the vanilla sale path.

Malkierian committed:
Missed some files for the pendingSale changes.

Also added Randomizer_Item_Give to the OnReceiveItem system.

Malkierian committed:
Ice traps now trigger OnReceiveItem.

Malkierian committed:
All ice traps truly do work now.

As a side effect, item autosave doesn't work for shop/scrub/merchant transactions, requires new OnSaleEnded hook that can also call the autosave.

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into item-give

Archez commented on soh/src/overlays/actors/ovl_En_GirlA/z_en_girla.c:
```diff
@@ -954,12 +987,18 @@ void EnGirlA_ItemGive_Randomizer(PlayState* play, EnGirlA* this) {
     }
 
+    if (getItemEntry.itemId == GI_ICE_TRAP || getItemEntry.itemId == RG_ICE_TRAP) {
+        GameInteractor_ExecuteOnReceiveItemHooks(ItemTable_RetrieveEntry(getItemEntry.modIndex, getItemEntry.getItemId));
```
Doesn't `ItemTable_RetrieveEntry()` return a GetItemEntry? Can we just use `getItemEntry` directly for the hook here?

Archez commented on soh/src/overlays/actors/ovl_En_Si/z_en_si.c:
```diff
@@ -123,6 +123,7 @@ void func_80AFB768(EnSi* this, PlayState* play) {
                         Audio_PlayFanfare_Rando(getItem);
                     } else {
                         gSaveContext.pendingIceTrapCount++;
+                        GameInteractor_ExecuteOnReceiveItemHooks(ItemTable_RetrieveEntry(getItem.modIndex, getItem.getItemId));
```
Same thing here, can we just pass in `getItem`?

Archez commented on soh/src/overlays/actors/ovl_En_Si/z_en_si.c:
```diff
@@ -161,6 +162,7 @@ void func_80AFB89C(EnSi* this, PlayState* play) {
                 Audio_PlayFanfare_Rando(getItem);
             } else {
                 gSaveContext.pendingIceTrapCount++;
+                GameInteractor_ExecuteOnReceiveItemHooks(ItemTable_RetrieveEntry(getItem.modIndex, getItem.getItemId));
```
Same thing here, can we just pass in `getItem`?

Archez commented on soh/src/code/z_sram.c:
```diff
@@ -108,6 +108,7 @@ void GiveLinksPocketItem() {
         } else if (getItemEntry.modIndex == MOD_RANDOMIZER) {
             if (getItemEntry.getItemId == RG_ICE_TRAP) {
                 gSaveContext.pendingIceTrapCount++;
+                GameInteractor_ExecuteOnReceiveItemHooks(ItemTable_RetrieveEntry(getItemEntry.modIndex, getItemEntry.getItemId));
```
Same thing here, can we just use `getIemEntry`?

Archez commented on soh/src/code/z_sram.c:
```diff
@@ -451,6 +452,7 @@ void Sram_InitSave(FileChooseContext* fileChooseCtx) {
             } else if (getItem.modIndex == MOD_RANDOMIZER) {
                 if (getItem.getItemId == RG_ICE_TRAP) {
                     gSaveContext.pendingIceTrapCount++;
+                    GameInteractor_ExecuteOnReceiveItemHooks(ItemTable_RetrieveEntry(getItem.modIndex, getItem.getItemId));
```
Same thing here, can we just use `getItem`?

Archez commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -6320,6 +6320,7 @@ s32 func_8083E5A8(Player* this, PlayState* play) {
                     Player_SetPendingFlag(this, play);
                     Message_StartTextbox(play, 0xF8, NULL);
                     Audio_PlayFanfare(NA_BGM_SMALL_ITEM_GET);
+                    GameInteractor_ExecuteOnReceiveItemHooks(ItemTable_RetrieveEntry(this->getItemEntry.modIndex, this->getItemEntry.getItemId));
```
Same thing here, can we just pass in this->getItemEntry directly? And all the ones below in this file.

Archez commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -12624,6 +12625,10 @@ s32 func_8084DFF4(PlayState* play, Player* this) {
             Player_SetPendingFlag(this, play);
         }
 
+        /*if (giEntry.itemId == RG_ICE_TRAP || giEntry.itemId == GI_ICE_TRAP) {
+            GameInteractor_ExecuteOnReceiveItemHooks(ItemTable_RetrieveEntry(giEntry.modIndex, giEntry.getItemId));
+        }*/
+
```
If this isn't needed, please remove it.

Archez reviewed (commented):
Doing a first pass through, main thing that stands out to me is a couple places where we can avoid a call to `ItemTable_ReceiveItem`

Malkierian commented:
Yeah, I feel pretty silly about all those unnecessary RetrieveEntry calls now, lol. No, there's no reason for them.

Malkierian committed:
Removed unnecessary calls to ItemTable_RetrieveEntry where GetItemEntry properties were being used to call it.  General code cleanup.

Malkierian committed:
Added OnSaleEnd hook for when rupees are finished deducting after a sale.

Migrated AutoSave to its own function, registered AutoSave function to OnReceiveItem and OnSaleEnd hooks to help with autsaving after buying items.

Some futureproofing for AutoSave function with parameters for skipping autosave, for when transition end is migrated to AutoSave function (whether through direct call or through a hook).

Malkierian commented:
Should be ready for merging pending another review.

dcvz commented on soh/soh/Enhancements/mods.cpp:
```diff
@@ -172,66 +172,74 @@ void RegisterSwitchAge() {
-            if ((CVarGetInteger("gAutosave", 0) == 2) || (CVarGetInteger("gAutosave", 0) == 5)) {
-                // Autosave for all items
-                Play_PerformSave(gPlayState);
+void AutoSave(GetItemEntry get) {
```
Let's name the param `itemEntry` so it's more descriptive.

dcvz commented on soh/soh/Enhancements/game-interactor/GameInteractor.h:
```diff
@@ -86,7 +87,8 @@ class GameInteractor {
     DEFINE_HOOK(OnExitGame, void(int32_t fileNum));
     DEFINE_HOOK(OnGameFrameUpdate, void());
-    DEFINE_HOOK(OnReceiveItem, void(uint8_t item));
+    DEFINE_HOOK(OnReceiveItem, void(GetItemEntry item));
```
Let's update the param names to `itemEntry` here and in the C exposed methods.

dcvz commented on soh/soh/Enhancements/mods.cpp:
```diff
@@ -172,66 +172,74 @@ void RegisterSwitchAge() {
+}
+
+void RegisterAutoSave() {
+    GameInteractor::Instance->RegisterGameHook<GameInteractor::OnReceiveItem>([](GetItemEntry get) { AutoSave(get); });
```
Same comment on param name being: `itemEntry`

dcvz reviewed (commented):
Looking good, just some small comments. I also don't see where you trigger the sale ended hooks, do we still need that?

Malkierian commented:
The call to `OnSaleEnd` happens on line 6198 of `z_parameter.c`, where the call used to be to `OnReceiveItem`.  But yeah, the param name change makes sense.

Malkierian commented:
Oh, and as a clarification, sale Item_Gives happen earlier than OnSaleEnd in other areas of code, but still utilize autosave limitations to prevent autosaving before rupee deduction.

Malkierian committed:
Renamed hook paramaters, and registered hook function parameters, to a more descriptive alternative.

dcvz commented on soh/soh/Enhancements/game-interactor/GameInteractor.h:
```diff
@@ -86,7 +87,8 @@ class GameInteractor {
     DEFINE_HOOK(OnExitGame, void(int32_t fileNum));
     DEFINE_HOOK(OnGameFrameUpdate, void());
-    DEFINE_HOOK(OnReceiveItem, void(uint8_t item));
+    DEFINE_HOOK(OnReceiveItem, void(GetItemEntry itemEntr));
```
Typo on entry :)

Malkierian replied:
Oh, good grief. I missed the exposures in `GameInteractor_Hooks.h` as well XD

Malkierian committed:
Missed a couple, fixed a typo.

Malkierian committed:
Merge branch 'develop' of https://github.com/HarbourMasters/Shipwright into item-give

Malkierian committed:
One more missed paramater name refactor.

Refactored all references to OnReceiveItem to OnItemReceive to mirror upcoming full hook refactor for name ordering conventions.

Up-to-date with develop.
